### PR TITLE
op-node: handle sequencer errors with derivation-levels, reset when L1 origins are inconsistent

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

On Goerli we have hit an interesting sequencer edge-case:
- The sequencer conf-depth is lower than the verifier conf-depth
- The L1 reorgs with a reorg-depth in-between these
- The verifier only ever sees the good chain
- The sequencer builds an unsafe L2 block with a L1 origin that doesn't match the canonical chain
- The verifier code-path drops the batch because of the bad L1 origin (epoch) value
- The verifier code-path patiently waits for the sequencer to provide an alternative L2 batch with good origin, because the sequencing window has not elapsed yet.
- The sequencer code-path doesn't undo the unsafe L2 block with bad L1 origin, nor does the verifier code-path

This situation locks up the chain until the sequencing window elapses. To prevent this, we need the sequencer to build a good L2 block, without having to restart it manually to get it back onto an unsafe L2 head with good L1 origin (the work-around we've executed two times now).

Changes:
- Removed unused `idleDerivation`; the sequencing and derivation codepaths are already independent, and handle conflicts through through the shared engine-control
- Added error-handling cases to the `RunNextSequencerAction` to handle when starting/sealing fails due to crit/reset/temp/other error.
- Changed the L1-origin consistency check in the sequencer to a reset-worthy error
- Make `RunNextSequencerAction` bubble up the error if it's critical, it's up to the driver to handle that (in this case log + shutdown driver, since we prefer safety over liveness).

**Tests**

- Added an action-test to run through this edge-case as regression test.
- Updated the sequencer chaos test to simulate these type of errors that require a `Reset()` call to recover from.

**Invariants**

Ensure that the selected L1 origin of every new sequenced L2 block is consistent (i.e. equal or parent-hash of new block matches current) with the previous L1 origin, and execute a reset to make things consistent otherwise.

**Additional context**

Follow-up to Goerli sequencing issue.

**Metadata**

Fix CLI-3404

### CODE REVIEW

File: op-e2e/actions/l2_sequencer.go [hunk 1]
```diff
@@ -103,7 +118,16 @@ func (s *L2Sequencer) ActBuildToL1Head(t Testing) {
 	}
 }
 
-// ActBuildToL1HeadExcl builds empty blocks until (excl.) the L1 head becomes the L2 origin
+// ActBuildToL1HeadUnsafe builds empty blocks until (incl.) the L1 head becomes the L1 origin of the L2 head
+func (s *L2Sequencer) ActBuildToL1HeadUnsafe(t Testing) {
+	for s.derivation.UnsafeL2Head().L1Origin.Number < s.l1State.L1Head().Number {
+		// Note: the
```
Comment (tynes) [inline op-e2e/actions/l2_sequencer.go:124]: looks line an unfinished comment here

File: op-node/rollup/derive/pipeline.go [hunk 2]
```diff
@@ -25,6 +25,11 @@ type L1Fetcher interface {
 	L1TransactionFetcher
 }
```
Comment (tynes) [inline op-node/rollup/derive/pipeline.go:27]: nit: godocs

File: op-node/rollup/driver/metered_engine.go [hunk 3]
```diff
@@ -31,9 +31,9 @@ type MeteredEngine struct {
 }
 
 // MeteredEngine implements derive.EngineControl
```
Comment (tynes) [inline op-node/rollup/driver/metered_engine.go:33]: nit: doc out of date

File: op-node/rollup/driver/sequencer.go [hunk 4]
```diff
@@ -159,28 +160,54 @@ func (d *Sequencer) BuildingOnto() eth.L2BlockRef {
 // RunNextSequencerAction starts new block building work, or seals existing work,
 // and is best timed by first awaiting the delay returned by PlanNextSequencerAction.
 // If a new block is successfully sealed, it will be returned for publishing, nil otherwise.
-func (d *Sequencer) RunNextSequencerAction(ctx context.Context) *eth.ExecutionPayload {
+// Only critical errors are bubbled up, other errors are handled internally.
+func (d *Sequencer) RunNextSequencerAction(ctx context.Context) (*eth.ExecutionPayload, error) {
 	if _, buildingID, _ := d.engine.BuildingPayload(); buildingID != (eth.PayloadID{}) {
 		payload, err := d.CompleteBuildingBlock(ctx)
 		if err != nil {
```
Comment (tynes) [inline op-node/rollup/driver/sequencer.go:167]: This function is very branchy, perhaps a switch statement will make it more readable. Its also nearly the same logic at the top and the bottom, perhaps some code could be deduplicated?
- adding time to `nextAction` could be turned into an internal function
- the loglines are the same, maybe a function with a callback where the logs happen internal to the function and the callback executes the actions that differ between the top branches? (`buildingID != (eth.PayloadId{})`?
- perhaps a top level `var` for null payload id to do comparisons against

Just suggestions if you like them

File: op-node/rollup/driver/state.go [hunk 5]
```diff
@@ -212,7 +209,11 @@ func (s *Driver) eventLoop() {
 
 		select {
 		case <-sequencerCh:
-			payload := s.sequencer.RunNextSequencerAction(ctx)
+			payload, err := s.sequencer.RunNextSequencerAction(ctx)
+			if err != nil {
+				s.log.Error("Sequencer critical error", "err", err)
```
Comment (tynes) [inline op-node/rollup/driver/state.go:214]: A metric here would be nice

File: op-node/rollup/driver/sequencer.go (hunk 4 above)
Comment (tynes) [reply op-node/rollup/driver/sequencer.go:167]: Also metrics on any types of errors could be useful here

File: op-e2e/actions/l2_sequencer_test.go [hunk 6]
```diff
@@ -100,3 +102,72 @@ func TestL2Sequencer_SequencerDrift(gt *testing.T) {
...
 	require.True(t, engine.l2ForceEmpty, "engine should not be allowed to include anything after sequencer drift is surpassed")
 }
+
+// TestL2Sequencer_SequencerOnlyReorg regression-tests a Goerli halt where the sequencer
+// would build an unsafe L2 block with a L1 origin that then gets reorged out,
+// while the verifier-codepath only ever sees the valid post-reorg L1 chain.
+func TestL2Sequencer_SequencerOnlyReorg(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	p := &e2eutils.TestParams{
+		MaxSequencerDrift:   20, // larger than L1 block time we simulate in this test (12)
```
Comment (tynes) [inline op-e2e/actions/l2_sequencer_test.go:112]: Is there any way to make an assertion on this in case the L1 block time gets changed in the future?

File: op-node/rollup/driver/sequencer.go (hunk 4 above)
Comment (protolambda) [reply op-node/rollup/driver/sequencer.go:167]: Unfortunately the error checking is using error-targets, not concrete values, so I can't use a switch statement.
And the log-lines are different; the messages indicate which branch we hit.
And I prefer an inline null payload-id check, since it's immutable and a default value.
Good point on metrics, will add those.

File: op-e2e/actions/l2_sequencer_test.go (hunk 6 above)
Comment (protolambda) [reply op-e2e/actions/l2_sequencer_test.go:112]: Will switch to default test params, this isn't strictly necessary, and the defaults should stay maintained as they are used in a lot of the tests.

File: op-node/rollup/driver/sequencer.go [hunk 7]
```diff
@@ -159,28 +169,57 @@ func (d *Sequencer) BuildingOnto() eth.L2BlockRef {
...
-			d.nextAction = d.timeNow().Add(time.Second)
-			if buildingID != (eth.PayloadID{}) { // don't keep stale block building jobs around, try to cancel them
-				d.CancelBuildingBlock(ctx)
+			if errors.Is(err, derive.ErrCritical) {
+				return nil, err // bubble up critical errors.
+			} else if errors.Is(err, derive.ErrReset) {
+				d.log.Error("sequencer failed to seal new block, requiring derivation reset", "err", err)
+				d.metrics.RecordSequencerReset()
+				d.nextAction = d.timeNow().Add(time.Second * time.Duration(d.config.BlockTime)) // hold off from sequencing for a full block
+				if buildingID != (eth.PayloadID{}) {                                            // cancel what we were doing
```
Comment (tynes) [inline op-node/rollup/driver/sequencer.go:183]: Do you need to call `engine.Reset` in this reset block?

Comment (tynes) [reply op-node/rollup/driver/sequencer.go:183]: Shouldn't the top check catch this condition? So it doesn't need to be checked here again? Would save some branching

File: op-node/rollup/driver/sequencer.go [hunk 8]
```diff
@@ -159,28 +169,57 @@ func (d *Sequencer) BuildingOnto() eth.L2BlockRef {
 // RunNextSequencerAction starts new block building work, or seals existing work,
 // and is best timed by first awaiting the delay returned by PlanNextSequencerAction.
 // If a new block is successfully sealed, it will be returned for publishing, nil otherwise.
-func (d *Sequencer) RunNextSequencerAction(ctx context.Context) *eth.ExecutionPayload {
+// Only critical errors are bubbled up, other errors are handled internally.
+func (d *Sequencer) RunNextSequencerAction(ctx context.Context) (*eth.ExecutionPayload, error) {
```
Comment (tynes) [inline op-node/rollup/driver/sequencer.go:173]: Would love some comments on expected invariants of this function. Its a really important function in the state machine

**Comment by protolambda** [comment]
@tynes ready for review again

**Review by tynes — APPROVED** [review]
